main/src/library/SetMultiMap.flix

```diff
@@ -0,0 +1,514 @@
+    ///
+    pub def memberOf(k: k, m: SetMultiMap[k, v]): Bool with Order[k] =
+        let SetMultiMap(m1) = m;
+        Map.memberOf(k, m1)
+
+    ///
+    /// Returns the keys of `m`.
+    ///
+    pub def keysOf(m: SetMultiMap[k, v]): Set[k] with Order[k] =
+        let SetMultiMap(m1) = m;
+        Map.keysOf(m1)
+
+    ///
+    /// Returns the values of `m`.
+    ///
+    /// Answer may contain duplicates where values were ascribed to multiple keys.
+    ///
+    pub def valuesOf(m: SetMultiMap[k, v]): List[v] with Order[k] =
+        let SetMultiMap(m1) = m;
+        Map.valuesOf(m1) |> List.map(Set.toList) |> List.flatten
+
+    ///
+    /// Updates `m` with `k => v`.
+    ///
+    pub def insert(k: k, v: v, m: SetMultiMap[k, v]): SetMultiMap[k, v] with Order[k], Order[v] =
+        let SetMultiMap(m1) = m;
+        SetMultiMap(Map.insertWith((v1, v2) -> Set.union(v1, v2), k, Set.singleton(v), m1))
+
+    ///
+    /// Updates `m` with `k => vs`.
+    ///
+    /// Where `vs` is any foldable container. If `vs` is empty nothing is inserted.
+    ///
+    pub def insertAll(k: k, vs: t[v], m: SetMultiMap[k, v]): SetMultiMap[k, v] with Order[k], Order[v], Foldable[t] =
+        match Foldable.toSet(vs) {
+            case s1 if Set.isEmpty(s1) => m
+            case s1                    => {
+                let SetMultiMap(m1) = m;
+                SetMultiMap(Map.insertWith((v1, v2) -> Set.union(v1, v2), k, s1, m1))
+            }
+        }
+
+    ///
+    /// Updates `m` with `k => map(f, vs)` if `k => vs` is in `m`.
+    ///
+    /// Otherwise, returns `m`.
+    ///
+    pub def adjust(f: v -> v & ef, k: k, m: SetMultiMap[k, v]): SetMultiMap[k, v] & ef with Order[k], Order[v] =
+        adjustWithKey((_, v1) -> f(v1), k, m)
+
+    ///
+    /// Updates `m` with `k => map(f(k), vs)` if `k => vs` is in `m`. Otherwise, returns `m`.
+    ///
+    pub def adjustWithKey(f: (k, v) -> v & ef, k: k, m: SetMultiMap[k, v]): SetMultiMap[k, v] & ef with Order[k], Order[v] =
+        updateWithKey((k1, v) -> Some(f(k1, v)), k, m)
+
+    ///
+    /// Updates `m` with `k => v1` if `k => v` is in `m` and `f(v) = Some(v1)`. Otherwise, returns `m`.
+    ///
+    pub def update(f: v -> Option[v] & ef, k: k, m: SetMultiMap[k, v]): SetMultiMap[k, v] & ef with Order[k], Order[v] =
+        updateWithKey((_, v1) -> f(v1), k, m)
+
+    ///
+    /// Updates `m` with `k => v1` if `k => v` is in `m` and `f(k, v) = Some(v1)`. Otherwise, returns `m`.
+    ///
+    pub def updateWithKey(f: (k, v) -> Option[v] & ef, k: k, m: SetMultiMap[k, v]): SetMultiMap[k, v] & ef with Order[k], Order[v] =
+        let SetMultiMap(m1) = m;
+        let outer = (k1, s) -> if (k1 == k) Set.map(v1 -> Option.getWithDefault(v1, f(k1, v1)), s) else s;
+        SetMultiMap(Map.mapWithKey(outer, m1))
+
+    ///
+    /// Removes the mapping `k` from the SetMultiMap `m`.
+    ///
+    pub def remove(k: k, m: SetMultiMap[k, v]): SetMultiMap[k, v] with Order[k] =
```

Comment:
What if I just want to remove one elm from the value set? What function should I use?

Reply:
Added `removeWithValue`.

The implement is not ideal as there is no function like `adjust` / `update` in `Map` and `RedBlackTree` to update-or-remove a value from a Map in one step (I don't want to leave keys pointing to empty ets).